mongo/integration/unified/client_entity.go

```diff
@@ -144,10 +159,30 @@ func (c *clientEntity) stopListeningForEvents() {
 	c.setRecordEvents(false)
 }
 
+func (c *clientEntity) isIgnoredEvent(event *event.CommandStartedEvent) bool {
+	// Check if command is not in ignoredCommands.
+	if _, ok := c.ignoredCommands[event.CommandName]; !ok {
+		// Check if command is "hello" or legacy hello.
+		if event.CommandName == "hello" || strings.ToLower(event.CommandName) == "ismaster" {
+			_, err := event.Command.LookupErr("speculativeAuthenticate")
+			speculativeAuth := err == nil
+
+			// If observeSensitiveCommands is false (or unset) and hello command is with
+			// speculative authenticate, command should be ignored.
+			if (c.observeSensitiveCommands == nil || !*c.observeSensitiveCommands) &&
+				speculativeAuth {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
 func (c *clientEntity) startedEvents() []*event.CommandStartedEvent {
 	var events []*event.CommandStartedEvent
 	for _, evt := range c.started {
-		if _, ok := c.ignoredCommands[evt.CommandName]; !ok {
+		if !c.isIgnoredEvent(evt) {
 			events = append(events, evt)
 		}
```

Comment:
Are we using `startedEvents()` to assert that an expected set of started events is published during a test? If so, why are we duplicating the command redaction logic in the test runner as well as in the driver? It seems more correct to adjust the expected set of started events in the test case, not redact them in the test runner.

Reply:
There are two places in our specifications where we redact security-sensitive commands.

In the [command monitoring spec](https://github.com/mongodb/specifications/blob/master/source/command-monitoring/command-monitoring.rst#security), we require that drivers redact security-sensitive commands and replies at the operation level. This prevents security-sensitive information from appearing in logs or test output due to command monitoring.

In the unified test format:

> ignoreCommandMonitoringEvents: Optional array of one or more strings. Command names for which the test runner MUST ignore any observed command monitoring events. The command(s) will be ignored in addition to configureFailPoint and any commands containing sensitive information (per the Command Monitoring spec) unless observeSensitiveCommands is true.

Writing tests and making sure that the expected set of events contains the correct auth-related commands can be problematic. Drivers differ on when certain auth-related commands are sent (particularly in setup), so a unified test runner that always monitors security-sensitive commands might produce different monitoring outputs depending on the driver and prevent a unified test.

Reply:
Cool, thanks for the detailed explanation!